### ISSUE TITLE
On the verification of descending heap sort

### ISSUE THREAD

**Issue by 0xD3ADBEEF** [OpenJML/OpenJML#718]
So, I've been experimenting with verification methods for heap sort, and unfortunately I ran into a lot of trouble. The key point of verifying heap sort is the `heapify` routine, which I implemented non-recursively as follows: 

```java
/*@
 	requires 1 <= i <= len <= arr.length;
 	requires \forall int k,j; 1 < k <= log2(len)-log2(i) && (i<<k)<= j <=((i+1)<<k) && j<= len;arr[j-1]>= arr[(j>>>1)-1];
 @*/
public static void heapify(/*@ non_null @*/ int [] arr, final int i, final int len) {
	int p = i;
	/*@ loop_invariant i <= p <= len;
	 	loop_invariant \forall int k,j; 1 < k <= log2(len)-log2(p) && (p<<k)<= j <=((p+1)<<k) && j<= len;arr[j-1]>= arr[(j>>>1)-1];
	 	decreasing len-p;
	@*/
	while(true){
		int c = p<<1; //Left child
		
		int m = p;
		if(i <= c && c <= len && arr[c-1] < arr[m-1]) {
			m = c;
		}
		
		c++; //Right child
		if(i <= c && c <= len && arr[c-1] < arr[m-1]) {
			m = c;
		}
		
		if(m != p) {				
			int tmp = arr[p-1];
			arr[p-1] = arr[m-1];
			arr[m-1] = tmp;
		} else {
			break;
		}
		//@ assert p == m>>1;
		p = m;
		//@ assert arr[(p>>1)-1] <= arr[p-1];
	}
}
```
Here, the prec. `requires \forall int k,j; 1 < k <= log2(len)-log2(i) && (i<<k)<= j <=((i+1)<<k) && j<= len;arr[j-1]>= arr[(j>>>1)-1];` basically says that the left and right child trees of node `i` are min. heaps. A similar loop invariant is also there, but in this case for node `p`. 

Now, here is where the actual issues kick in. The routine `log2` can be naively implemented as `Math.log(v)/Math.log(2)`. Unfortunately, this implementation results in an error from the tool. I read somewhere on SO and in bug reports that this is because it uses doubles, and the tool cannot handle these pretty well yet.

Fair enough, I implemented a naive integer version as follows:
```java
/*@ 
 	requires 0 < i;
 	ensures (\old(i) >> \result) == 1;
 @*/
public static final /*@ pure @*/ int log2(int i) {
	int log = 0;
	/*@
	 decreasing i-1;
	 loop_invariant i == (\old(i)>>log);
	 loop_invariant 1 <= i <= \old(i);
	 @*/
	while(i > 1) {
		i = i>>1;
		log++;
	}
	return log;
}
```
This is verified and works. However, using this routine in the invariants causes the tool to hang. I dunno if this is because of the bit arithmetic, but a second guess would be that it is a costly routine, given that it uses a loop, so I implemented a "fast" version as follows:
```java
/*@ 
 	requires 0 < i;
 	ensures (\old(i) >> \result) == 1;
 @*/
public static pure model int log2( int i ) // returns 0 for bits=0
{
    int log = 0;
    if( (i >>> 16) != 0 ) { log = 16; i = i >>> 16; }
    if( i >= 256 ) { i = i >>> 8; log += 8; }
    if( i >= 16  ) { i = i >>> 4; log += 4; }
    if( i >= 4   ) { i = i >>> 2; log += 2; }
    return log + ( i >>> 1 );
}
```
This is also verified, and works, but guess what: same problem, the tool hangs :) 

Okay, fine. Then I went on for a more pragmatic approach, and replaced the prec., and similarly the loop invariant, with the following:
``` java
requires \forall int k; (i<<2) <= k <= len && isRoot(i<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Left child tree is heap
requires \forall int k; (i<<2) <= k <= len && isRoot((i<<1) + 1,k); arr[k-1] >= arr[(k>>>1)-1]; //Right child tree is heap
```
where `isRoot(i,n)` tests if `i` is the root of a subtree that contains node `n`. The first obvious implementation of this would be `\exists int j;1 < j;(n>>>j)==i`, but this, unfortunately, doesn't work, as this simple test:
```java
int i = 4, n = 16;		
//@ assert \exists int j;1 < j;(n>>>j)==i;
```
fails to establish the assertion. Even this stronger test fails:
```java
int i = 4, n = 16;		
//@ assert (n>>>2)==i;
//@ assert \exists int j;1 < j <= Integer.MAX_VALUE;(n>>>j)==i;
```
So, I implemented a recursive version of `isRoot` with division and multiplication:
```java
/*@
 requires 1 <= i <= n;
 ensures (n < (i<<1)) ==> \result == (i==n);
 ensures (n >= (i<<1)) ==> \result == isRoot(i,n>>>1);
 ensures ((i<<1) < 0) ==> !\result;
 @*/
public static /*@ pure @*/ boolean isRoot(final int i, final int n) {
	if((i<<1) > 0) {
		if(n >= (i<<1)) {
			return isRoot(i,n>>>1);
		}
		return i==n;
	}
	return false;
}
```
I'm not sure if my preconditions and postconditions make sense, but this works, and is verified. But, when I run the simple test
```java
//@ assert isRoot(1,Integer.MAX_VALUE);
//@ assert \forall int k; 1 < k <= Integer.MAX_VALUE; isRoot(1,k);
```
then the assertions are invalid. But, it is fairly easy to verify this exhaustively by running `isRoot` over all integers in the range `(1,Integer.MAX_VALUE]`:
```java
boolean works = true;
System.out.println("Running test...");
for(int i = 2;i <= Integer.MAX_VALUE && 0 <= i;i++) {
	works = HeapSort.isRoot(1,i);
	if(!works) {
		System.out.println("CE: " + i);	
		break;
	}
}
if(works) {
	System.out.println("Yes, it works");
}
```
No problem, this following implementation using `log2`:
```java
/*@ 
 	requires 1 <= i <= n;
 	ensures log2(n)-log2(i) < 1 ==> !\result;
 	ensures log2(n)-log2(i) >= 1 ==> \result == ((n>>>(log2(n)-log2(i))) == i);
 @*/
public static /*@ pure @*/ boolean isRoot(final int i, final int n) {
	int d = log2(n)-log2(i);
	if(d >= 1) {
		return (n>>>d) == i;
	}
	return false;
}
```
Passes the above tests and is verified. But, it hangs when used to verify `heapify`.

Maybe I'm just impatient and don't let the tool run long enough to give me a constructive answer. Or, maybe the tool is still not mature enough to handle this class of assertions. I dunno, but this really demotivates me from using it. 

Any ideas and/or code examples that show how to properly verify this type of methods? And are the discussed issues bugs, or just limitations of the tool?

**Comment by davidcok** [comment]
The principal problem is that both routines are being proved using bit-vector arithmetic, and that takes MUCH longer than proving equivalent problems with integer arithmetic. One way to help is to limit the size of the bit-vectors; for example, at least for the purposes of debugging the invariants, use arr.length <= 64.

Also, though this should not be a problem, I would rewrite log2 to avoid using \old and avoid changing the formal parameter.

**Comment by 0xD3ADBEEF** [comment]
> The principal problem is that both routines are being proved using bit-vector arithmetic, and that takes MUCH longer than proving equivalent problems with integer arithmetic. One way to help is to limit the size of the bit-vectors; for example, at least for the purposes of debugging the invariants, use arr.length <= 64.

That is unfortunate, but I agree that for now it is best that I restrict the length to something small. Curious, is it because internally the tool iterates over every bit?

I'm also trying an alternative method by making a lookup table as follows:
```java
public final /*@ non_null \readonly static_initializer@*/static int [] binarr = new int[]
	{0x1,0x2,0x4,0x8,
	0x10,0x20,0x40,0x80,
	0x100,0x200,0x400,0x800,
	0x1000,0x2000,0x4000,0x8000,
	0x10000,0x20000,0x40000,0x80000,
	0x100000,0x200000,0x400000,0x800000,
	0x1000000,0x2000000,0x4000000,0x8000000,
	0x10000000,0x20000000,0x40000000
};
//@ public invariant  binarr[0]==1 && \forall int j;1 <= j < 31;binarr[j]==2*binarr[j-1];
```
so then the arithmetic can be skipped entirely, however I am getting the following error from the type checker: An identifier with package visibility may not be used in a invariant clause with public visibility.

**Comment by davidcok** [comment]
Skipping the shift and log2 arithmetic would help greatly — especially as the log2 function is only declared implicitly.

Regarding An identifier with package visibility may not be used in a invariant clause with public visibility
Either declare the identifier public or //@ spec_public

**Comment by 0xD3ADBEEF** [comment]
> Skipping the shift and log2 arithmetic would help greatly — especially as the log2 function is only declared implicitly.

I'll report back on my findings.

> Regarding An identifier with package visibility may not be used in a invariant clause with public visibility Either declare the identifier public or //@ spec_public

The identifier is already declared public, unless I messed up the syntax. However, removing the `public` keyword in front of `invariant` silences the error. Not sure if that is intended behaviour, since the LUT array is static. Nevertheless, running ESC gives some weird indexing errors, so I added an extra invariant, totalling to
```java
//@ static invariant binarr.length > 0 && binarr.length==31;
//@ static invariant binarr[0]==1 && \forall int j;1 <= j < binarr.length;binarr[j]==2*binarr[j-1];
```
Some initial tests:
```java
//@ assert \exists int j;0<=j< binarr.length;500/binarr[j]==1;
```
Does not work, unless I supplement it with a preceding
```java
//@ assert \exists int j;0<=j< binarr.length-1;binarr[j]<=500 && binarr[j+1] > 500;
```
Trying to implement `log2` as:
```java
//@ assert (\max int j;0<=j< binarr.length && binarr[j]<=500;j)==8;
```
does not seem to work either (complains about max not being supported?). In the trace it gives the wrong answer 14:
```
/home/viktorio/eclipse-workspace/Learning/src/learning/sorting/HeapSort.java:71:  	assert Assert \exists int j; 0 <= j < binarr.length - 1; binarr[j] <= 500 && binarr[j + 1] > 500; ...
			VALUE: \exists int j; 0 <= j < binarr.length - 1; binarr[j] <= 500 && binarr[j + 1] > 500	 === true
/home/viktorio/eclipse-workspace/Learning/src/learning/sorting/HeapSort.java:72:  	assert Assert \exists int j; 0 <= j < binarr.length; 500 / binarr[j] == 1;
			VALUE: \exists int j; 0 <= j < binarr.length; 500 / binarr[j] == 1	 === true
/home/viktorio/eclipse-workspace/Learning/src/learning/sorting/HeapSort.java:73:  	assert Assert (\max int j; 0 <= j < binarr.length && binarr[j] <= 500; j) == 8;
			VALUE: \max int j; 0 <= j < binarr.length && binarr[j] <= 500; j	 === 14
			VALUE: (\max int j; 0 <= j < binarr.length && binarr[j] <= 500; j)	 === 14
			VALUE: 8	 === 8
			VALUE: (\max int j; 0 <= j < binarr.length && binarr[j] <= 500; j) == 8	 === false
/home/viktorio/eclipse-workspace/Learning/src/learning/sorting/HeapSort.java:73:  Invalid assertion (Assert)
```

**Comment by davidcok** [comment]
I would be surprised in the invariant did not need to be public, at least if the methods are public

**Comment by davidcok** [comment]
Give me a full class if you want some help in debugging or at least comments on your error messages

**Comment by 0xD3ADBEEF** [comment]
It's a huge mess, and I commented out a lot of things. This is my work so far:
```java
public class HeapSort {	
	public final /*@ non_null \readonly static_initializer@*/ static int [] binarr = new int[]
		{0x1,0x2,0x4,0x8,
		0x10,0x20,0x40,0x80,
		0x100,0x200,0x400,0x800,
		0x1000,0x2000,0x4000,0x8000,
		0x10000,0x20000,0x40000,0x80000,
		0x100000,0x200000,0x400000,0x800000,
		0x1000000,0x2000000,0x4000000,0x8000000,
		0x10000000,0x20000000,0x40000000
	};
	//@ static invariant binarr.length > 0 && binarr.length==31;
	//@ static invariant binarr[0]==1 && \forall int j;1 <= j < binarr.length;binarr[j]==2*binarr[j-1];
	
	
	
//	/*@
//	 requires 1 <= i <= len <= arr.length;
//	 requires \forall int k; (i<<2) <= k <= len && isRoot(i<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Left child tree is heap
//	 requires \forall int k; (i<<2) <= k <= len && isRoot((i<<1) + 1,k); arr[k-1] >= arr[(k>>>1)-1]; //Right child tree is heap
//	 @*/
//	public static void heapify(/*@ non_null @*/ int [] arr, final int i, final int len) {
//		int c = i<<1, p=i; //Left child
//		if(0 < c && c <= len) {
//			if(arr[c-1] < arr[i-1])
//				p=c;
//		}
//		c++; //Right child
//		if(0 < c && c <= len) {
//			if(arr[c-1] < arr[i-1])
//				p=c;
//		}
//		//@ assert (p!=i) ==> \forall int k; (p<<1) <= k <= len && isRoot(i<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Tree of p is heap
//		//@ assert (p!=i) ==> \forall int k; (p<<2) <= k <= len && isRoot(p<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Left child tree is heap
//		//@ assert (p!=i) ==> \forall int k; (p<<2) <= k <= len && isRoot((p<<1) + 1,k); arr[k-1] >= arr[(k>>>1)-1]; //Right child tree is heap
////		if(p!=i) {
////			int tmp=arr[i-1];
////			arr[i-1]=arr[p-1];
////			arr[p-1]=tmp;
////			//@ assert arr[i-1] <= arr[p-1];
////			heapify(arr,p,len);
////		}
//	}
	

//	/*@
//	requires 0 < i;
//	ensures (\old(i) >>> \result) == 1;
//	@*/
//	public static /*@ pure @*/ int log2( int i ) // returns 0 for bits=0
//	{
//	    int log = 0;
//	    if( (i >>> 16) != 0 ) { log = 16; i = i >>> 16; }
//	    if( i >= 256 ) { i = i >>> 8; log += 8; }
//	    if( i >= 16  ) { i = i >>> 4; log += 4; }
//	    if( i >= 4   ) { i = i >>> 2; log += 2; }
//	    return log + ( i >>> 1 );
//	}
//	
	public static void test() {
		//int [] arr = {1,2,3,41,56,6,8,111,56};
		//int i = 4, n = 16;		
		// assert (n>>>2)==i;
		// assert \exists int j;1 < j <= Integer.MAX_VALUE;(n>>>j)==i;
		
		//@ assert \exists int j;0<=j< binarr.length-1;binarr[j]<=500 && binarr[j+1] > 500;
		//@ assert \exists int j;0<=j< binarr.length;500/binarr[j]==1;
		//@ assert (\max int j;0<=j< binarr.length && binarr[j]<=500;j)==8;
	}
	
	
	/*@
	 	requires 1<=i && 0 <= n;
	 	ensures \result <==> i==n/2;
	 @*/
//	public static /*@ pure @*/ final boolean isChild(final int i, final int n) {
//		return i == n/2;
//	}
//	
//	/*@
//	 	requires 1<=i && 2*i <= n <= Integer.MAX_VALUE;
//	 @*/
//	public static /*@ pure @*/ final boolean isRoot(final int i, int n) {
//		while(2*i <= n/2) {
//			
//		}
//		return n/2 == i;
//	}
//	/*@
//	 requires 1 <= i <=  n <= Integer.MAX_VALUE;
//	 ensures 1 <= n <= \old(n) && 1 <= n <= i;
//	 ensures \result <==> (n==i);
//	 @*/
//	public static /*@ pure @*/ boolean isRoot(final int i, final int n) {
//		/*@
//		 loop_modifies n;
//		 decreasing n;
//		 loop_invariant 1 <= n <= \old(n);
//		 loop_invariant i/2 <= n;
//		 @*/
//		while(n > i) {
//			n = n / 2;
//		}
//		//@ assert 1 <= n <= i;
//		return n==i;
//	}
	
//	/*@ 
//	 	requires 1 <= i <= n;
//	 	ensures log2(n)-log2(i) < 1 ==> !\result;
//	 	ensures log2(n)-log2(i) >= 1 ==> \result == ((n>>>(log2(n)-log2(i))) == i);
//	 @*/
//	public static /*@ pure @*/ boolean isRoot(final int i, final int n) {
//		int d = log2(n)-log2(i);
//		if(d >= 1) {
//			return (n>>>d) == i;
//		}
//		return false;
//	}

		
//	/*@ 
//	 	requires 0 < i;
//	 	ensures (\old(i) >> \result) == 1;
//	 @*/
//	public static final /*@ pure @*/ int log2i(int i) {
//		int log = 0;
//		/*@
//		 decreasing i-1;
//		 loop_invariant i == (\old(i)>>log);
//		 loop_invariant 1 <= i <= \old(i);
//		 @*/
//		while(i > 1) {
//			i = i>>1;
//			log++;
//		}
//		return log;
//	}
	
//	/*@
//	 	requires 1 <= i <= len <= arr.length;
//	 	//requires \forall int k,j; 1 < k <= log2(len)-log2(i) && (i<<k)<= j <=((i+1)<<k) && j<= len;arr[j-1]>= arr[(j>>>1)-1];
//	 	requires \forall int k; (i<<2) <= k <= len && isRoot(i<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Left child tree is heap
//	 	requires \forall int k; (i<<2) <= k <= len && isRoot((i<<1) + 1,k); arr[k-1] >= arr[(k>>>1)-1]; //Right child tree is heap
//	 @*/
//	public static void heapify(/*@ non_null @*/ int [] arr, final int i, final int len) {
//		int p = i;
//		/*@ loop_invariant i <= p <= len;
//		 	//loop_invariant \forall int k,j; 1 < k <= log2(len)-log2(p) && (p<<k)<= j <=((p+1)<<k) && j<= len;arr[j-1]>= arr[(j>>>1)-1];
//		 	loop_invariant \forall int k; (p<<2) <= k <= len && isRoot(p<<1,k); arr[k-1] >= arr[(k>>>1)-1]; //Left child tree is heap
//	 		loop_invariant \forall int k; (p<<2) <= k <= len && isRoot((p<<1) + 1,k); arr[k-1] >= arr[(k>>>1)-1]; //Right child tree is heap
//		 	decreasing len-p;
//		@*/
//		while(true){
//			int c = p<<1; //Left child
//			
//			int m = p;
//			if(i <= c && c <= len && arr[c-1] < arr[m-1]) {
//				m = c;
//			}
//			
//			c++; //Right child
//			if(i <= c && c <= len && arr[c-1] < arr[m-1]) {
//				m = c;
//			}
//			
//			if(m != p) {				
//				int tmp = arr[p-1];
//				arr[p-1] = arr[m-1];
//				arr[m-1] = tmp;
//			} else {
//				break;
//			}
//			//@ assert p == m>>1;
//			p = m;
//			//@ assert arr[(p>>1)-1] <= arr[p-1];
//		}
//	}
	
//	public static void buildHeap(/*@ non_null @*/ int [] arr) {
//		for(int i = arr.length>>1;i > 0;i--) {
//			heapify(arr,i,arr.length);
//		}
//	}
}
```

**Comment by 0xD3ADBEEF** [comment]
Does JML understand recursion? I have the following method:
```java
/*@
 requires 1 <= i <= len <= arr.length <= Integer.MAX_VALUE;
 ensures i > len/2 ==> \result;
 ensures (i <= len/2 && arr[2*i-1] < arr[i-1]) ==> !\result;
 ensures (i <= len/2 && arr[2*i-1] >= arr[i-1] && i <= (len-1)/2 && arr[2*i] < arr[i-1]) ==> !\result;
 ensures (i <= len/2 && arr[2*i-1] >= arr[i-1] && i > (len-1)/2 ) ==> (\result <==> isMinHeap(arr,2*i,len));
 ensures (i <= len/2 && arr[2*i-1] >= arr[i-1] && i <= (len-1)/2 && arr[2*i] >= arr[i-1]) ==> (\result <==> (isMinHeap(arr,2*i,len) && isMinHeap(arr,2*i+1,len)));
 @*/
static /*@ pure @*/boolean isMinHeap(final /*@ non_null \readonly @*/ int [] arr, final int i, final int len) {
	boolean lHeap = true, rHeap = true;
	if(i <= len/2) {
		//@ assert 1 <= 2*i <= len;
		int c = 2*i; //Left child
		if(arr[c-1] < arr[i-1]) {
			return false; //child is smaller, so not heap
		} 
		lHeap = isMinHeap(arr,c,len);
		if(i <= (len-1)/2) {
			//@ assert 1 <= 2*i+1 <= len;
			c++; //Right child
			if(arr[c-1] < arr[i-1]) {
				return false; //child is smaller, so not heap
			} 
			rHeap = isMinHeap(arr,c,len);
		}
	}
	return lHeap && rHeap; //i is heap
}
```
and the following code is verified:
```java
public static void test() {
	int [] arr = {1,2,3,41,56,6,8,111,56};
	int i = 1, len = 7;
	
	//@ assert 1 <= i <= len <= arr.length <= Integer.MAX_VALUE;
	//@ assert isMinHeap(arr,4*i,len);
	//@ assert isMinHeap(arr,4*i+1,len);
	//@ assert isMinHeap(arr,4*i+2,len);
	//@ assert isMinHeap(arr,4*i+3,len);
	
	//@ assert isMinHeap(arr,2*i,len);
	//@ assert isMinHeap(arr,2*i+1,len);
	boolean result1 = isMinHeap(arr,i,len);
	//@ assert result1;
	//@ assert isMinHeap(arr,i,len);
}
```
**only** if I have the block of asserts
```java
//@ assert 1 <= i <= len <= arr.length <= Integer.MAX_VALUE;
//@ assert isMinHeap(arr,4*i,len);
//@ assert isMinHeap(arr,4*i+1,len);
//@ assert isMinHeap(arr,4*i+2,len);
//@ assert isMinHeap(arr,4*i+3,len);
//@ assert isMinHeap(arr,2*i,len);
//@ assert isMinHeap(arr,2*i+1,len);
```
Seems to me that I can't force the checker to work with recursive specifications. Is my specification of `isMinHeap` even correct?

**Comment by 0xD3ADBEEF** [comment]
David,

After much trial and error, I have managed to verify heap sort. The key was to relax the preconditions a little bit:
```java
/*@ requires 1 <= i <= len/2 && len <= arr.length;
 	//New spec
 	requires \forall int k;i < k <= len/2;arr[k-1] <= arr[2*k-1]; 
 	requires \forall int k;i < k <= (len-1)/2;arr[k-1] <= arr[2*k];
	ensures \forall int k;i <= k <= len/2;arr[k-1] <= arr[2*k-1];
 	ensures \forall int k;i <= k <= (len-1)/2;arr[k-1] <= arr[2*k]; 
 	
 	//Old spec
 	//requires \forall int k;2*i < k-1 <= len-1;arr[k-1] >= arr[k/2-1];	
 	//ensures  \forall int k;2*i <= k <= len;arr[k-1] >= arr[k/2-1];
 	//ensures  arr[i-1] <= arr[2*i-1];
 	//ensures  2*i <= arr.length-1 ==> arr[i-1] <= arr[2*i];
 @*/
public static void heapify(/*@ non_null @*/ int [] arr, final int i, final int len) {
	int j = i;
	
	/*@ loop_invariant i <= j <= len;
	 	//New loop invs.
	 	loop_invariant \forall int k;j < k <= len/2;arr[k-1] <= arr[2*k-1];
 		loop_invariant \forall int k;j < k <= (len-1)/2;arr[k-1] <= arr[2*k];
 		loop_invariant \forall int k;i <= k < j && k <= len/2;arr[k-1] <= arr[2*k-1];
 		loop_invariant \forall int k;i <= k < j && k <= (len-1)/2;arr[k-1] <= arr[2*k];
 		
 		//Old loop invs.
	 	//loop_invariant \forall int k;2*j < k-1 <= len-1;arr[k-1] >= arr[k/2-1];	
	 	//loop_invariant \forall int k; 2*i <= k < 2*j && k <= len;arr[k-1] >= arr[k/2-1]; 
		loop_invariant \forall int k; 2*j-1<=k-1<=2*j && k <= len && j > i;arr[k-1] >= arr[j/2-1]; 
	 	decreasing len-j;
	 @*/
	while(true) {			
		int m = j;
		if(m <= len/2) {
			int c = 2*m;
			if(arr[c-1] < arr[m-1])
				m=c;
			if(c < len && arr[c] < arr[m-1])
				m=c+1;
		}
		
		if(m==j) {
			break;
		}
		int tmp = arr[j-1];
		arr[j-1] = arr[m-1];
		arr[m-1] = tmp;
		j = m;
	}
}

/*@
 requires 1 <= len <= arr.length;
 requires \forall int k;1 <= k <= len/2;arr[k-1] <= arr[2*k-1];
 requires \forall int k;1 <= k <= (len-1)/2;arr[k-1] <= arr[2*k]; 
 ensures  \result && \forall int k;0 <= k < len;arr[k] >= arr[0]; //Always true
public static pure model boolean isGeq(final  non_null  int [] arr, final int len) {
	  loop_invariant 0 <= i < len;
	  loop_invariant \forall int k;i < k < len;arr[k] >= arr[0];
	  decreasing i;
	for(int i = len-1;i > 0;i--) {
		int j=i+1;
		
		loop_invariant 1 <= j <= i+1;
		loop_invariant arr[i] >= arr[j-1];
		decreasing j;
		while(j > 1) {
			j = j / 2;
		}
	}
	return \forall int k;0 < k < len;arr[k] >= arr[0];
}@*/

/*@
 ensures \forall int k,j; 0 <= k < j < arr.length;arr[k] >= arr[j];
 @*/
public static void sort(/*@ non_null @*/ int [] arr) {
	//Array of size 1 is already sorted
	if(arr.length < 2)
		return;
		
	//Build heap
	/*@ loop_invariant 0 <= i <= arr.length/2;
	 	loop_invariant \forall int k;i < k <= arr.length/2;arr[k-1] <= arr[2*k-1];
 		loop_invariant \forall int k;i < k <= (arr.length-1)/2;arr[k-1] <= arr[2*k]; 
	 	//loop_invariant \forall int k;2*i < k-1 <= arr.length-1;arr[k-1] >= arr[k/2-1];
	 	decreasing i;
	 @*/
	for(int i = arr.length/2;i > 0;i--) {
		heapify(arr,i,arr.length);
	}
	
	//Now sort
	int tmp;
	/*@ loop_invariant 1 <= len < arr.length;
	 	loop_invariant \forall int k;1 <= k <= (len+1)/2;arr[k-1] <= arr[2*k-1];
 		loop_invariant \forall int k;1 <= k <= len/2;arr[k-1] <= arr[2*k]; 
 		loop_invariant isGeq(arr,len+1);
 		loop_invariant \forall int k,j; 0 <= k <= len < j < arr.length;arr[j] <= arr[k];
 		loop_invariant \forall int k,j; len <= k < j < arr.length;arr[j] <= arr[k];
	 	decreasing len-1;
	 @*/
	for(int len = arr.length-1;len>1;len--) {
		tmp = arr[0];
		arr[0] = arr[len];
		arr[len] = tmp;
		heapify(arr,1,len);
	}
	tmp = arr[0];
	arr[0] = arr[1];
	arr[1] = tmp;
}
```
I had an issue with asserting `\forall int k;0 <= k < len;arr[k] >= arr[0];` after the initial heap is built, i.e. after both `\forall int k;1 <= k <= len/2;arr[k-1] <= arr[2*k-1];` and `\forall int k;1 <= k <= (len-1)/2;arr[k-1] <= arr[2*k];` hold true, which is the reason I included the dummy model method `isGeq`. The method also resolves another weird issue, where the checker does not distinguish between a 1-indexed `arr[i-1]` and 0-indexed `arr[i]`.

To summarize some the other issues/bugs that I experienced throughout this whole endeavour:

1. Math functions involving doubles do not work.
2. Bit-shift arithmetic is painfully slow.
3. Declaring class invariants on public static final arrays does not work, even after using the clause `static_initializer` on the array.
4. Declaring a non-static final array `binarr` with invariants works, but using it as `(new HeapSort).binarr[i]` for some `i` does not work, and gives an error.
5. Declaring recursive specs of a method using the method itself works to a limited level.

Happy new year!

**Comment by davidcok** [comment]
Viktorio, Thanks for the follow-up

1) May I include this (perhaps edited) as a public example of OpenJML?

2) To your points:
a) Even if double arithmetic were better specified, it would suffer from slow performance because of the solver
b) bit-vector arithmetic is also generally a solver performance problem
c) On 3-4-5, I would be glad for example code that you thought should work but does not. Its may just be a case of an incomplete implementation.

- David

**Comment by 0xD3ADBEEF** [comment]
> 1) May I include this (perhaps edited) as a public example of OpenJML?

Definitely! And, would greatly appreciate improvements! :)

> a) Even if double arithmetic were better specified, it would suffer from slow performance because of the solver

Agreed, unless perhaps some form of abstraction/overapproximation is used.

> b) bit-vector arithmetic is also generally a solver performance problem

Yeah, I imagine there's a lot going on under the hood. I don't think improving performance would be sufficient, but who knows.

> c) On 3-4-5, I would be glad for example code that you thought should work but does not. Its may just be a case of an incomplete implementation.

I hope the examples in my previous comments demonstrate these points sufficiently. Otherwise I we can isolate them and post in a separate comment?

**Comment by davidcok** [comment]
Just email me directly with any examples that fail and demonstrate the deficiencies that you point out.

**Comment by 0xD3ADBEEF** [comment]
Which email should I use? Also, would you like other verified sorting examples (bubble, selection, merge, etc..)? Of course, I would appreciate that I be credited somewhere in the source code :)

**Comment by davidcok** [comment]
Send to this one: David.r.cok@gmail com

And of course I will add something like “Contributed by … “ — please provide me a full name.

**Comment by davidcok** [comment]
Regarding the recursive e countDown example above. It appears that this is a limitation of the solver -- it must unroll the recursive definition only a limited humber of times.

More typically, either (1) the specifications for countDown would "solve" the recursion as in
/*@
 requires c >= 0;
 ensures \result == 0;
 @*/
static /*@ pure @*/int countDownA(final int c) {
        if(c > 0)
                return countDownA(c-1);
        return 0;
}

which the tool can prove, or (2) at the call site, one only needs a few recursive unfoldings to do the proof.

**Comment by 0xD3ADBEEF** [comment]
I suspected just as much, and it appears that the maximum number of unrolls is 21. Interesting that the above version is proven. Sent you an email by the way.

**Comment by davidcok** [comment]
The test cases generated from this issue now work. 
Also HeapSort has been made a demo example (and a test case for it also): https://github.com/OpenJML/openjml.github.io/tree/master/examples

Leaving the issue open for consideration of other thoughts in the discussion.
